### PR TITLE
walletInteraction flag

### PR DESCRIPTION
**Pull request by kushti**
In this PR:

* Hybrid database option has been removed
* *walletInteraction* flag added to a scan. Flag values: 
  - off - if a box associated with a scan with interaction flag == off, the box will be added to the scan only
  - shared - box can be in both the wallet and the scan if wallet finds it (box is protected by wallet-related P2PK)
  - forced - box is always added to the wallet if it is added to the scan

### CODE REVIEW

File: src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala [hunk 1]
```diff
@@ -15,54 +18,72 @@ import scala.util.{Failure, Success, Try}
   * @param scanId         - unique identifier of an scan in the local system
   * @param scanName       - scan description (255 bytes in UTF-8 encoding max)
   * @param trackingRule  - a predicate to scan the blockchain for specific scan-related boxes
+  * @param walletInteraction - a flag which is prescribing how the scan is interacting with the p2pk-wallet
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala:21]: Let's not call it a "flag" and use "type" or "kind" instead. Flag is usually binary (0 or 1), and here it is ternary and potentially even more.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala:21]: @aslesarenko flag is not necessarily a binary value:  https://techterms.com/definition/flag . 

Both terms you proposed are not so good here imho.

File: src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala [hunk 2]
```diff
@@ -277,12 +277,12 @@ class WalletRegistry(store: LDBVersionedStore)(ws: WalletSettings) extends Score
...
     *
     * @param newScans - ids of new scans box should be associated with
-    * @param box - box to be updated
+    * @param box - box to be updated (new version)
     * @return
     */
   def updateScans(newScans: Set[ScanId], box: ErgoBox): Try[Unit] = Try {
     val bag0 = KeyValuePairsBag.empty
-    val oldBox = getBox(box.id)
+    val oldBox = getBox(box.id) // read old version from the database
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/persistence/WalletRegistry.scala:285]: The same id as it is the same box, but with different scans. Already reflected in ScalaDoc

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala [hunk 3]
```diff
@@ -218,37 +219,58 @@ object WalletScanLogic extends ScorexLogging {
     val externalScans: Seq[Scan] = walletVars.externalScans
 
     tx.outputs.flatMap { bx =>
-      val appsTriggered = externalScans.filter(_.trackingRule.filter(bx)).map(app => app.scanId)
+
+      // First, we check apps triggered by the tx output
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala:223]: @aslesarenko the question is vague, what kind of answer are you looking for ?

File: src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala [hunk 4]
```diff
@@ -15,54 +18,72 @@ import scala.util.{Failure, Success, Try}
...
+  }
+
+  override def parse(r: Reader): Scan = {
+    val scanId = ScanId @@ r.getShort()
+    val appName = r.getShortString()
+    val pos = r.position
+
+    // hack to read scans serialized with previous versions (they will have positive first byte)
+    // for scans written with previous versions, walletInteraction flag is set to "off"
+    val interactionFlag = r.getByte() match {
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala:49]: In one case (out of two) we still need to advance the position, so what's the advantage of peekByte then ?

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass !

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala:223]: I don't know, I just don't understand this comment.

File: src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala (hunk 1 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala:21]: I don't want to argue here, just a suggestion. 
From the definition "Flags are often binary flags" is kind of what I wrote above "Flag is usually binary".

File: src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala:49]: How do you know you are working with position in a correct way?
Do you know exactly how setting of position is implemented in the actual reader class passed to this method?
The peekByte and getByte are semantically less questionable.
You can call getByte conditionally in the branch as needed, and can also ignore the returned value.

[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/scanning/Scan.scala]: ok done

File: src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala:223]: ok and what to do?

**Comment by kushti** [comment]
@aslesarenko comments addressed, please make another pass!

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**